docs/src/userguide/language_basics.rst

```diff
@@ -581,44 +906,93 @@ be careful what you do.
-Here is an example:
+        .. code-block:: cython
 
-.. literalinclude:: ../../examples/userguide/language_basics/casting_python.pyx
+            cdef char *p
+            cdef float *q
+            p = <char*>q
 
-The precedence of ``<...>`` is such that ``<type>a.b.c`` is interpreted as ``<type>(a.b.c)``.
+        When casting a C value to a Python object type or vice versa,
+        Cython will attempt a coercion. Simple examples are casts like ``<int>pyobj``,
+        which converts a Python number to a plain C ``int`` value, or ``<bytes>charptr``,
+        which copies a C ``char*`` string into a new Python bytes object.
 
-Casting to ``<object>`` creates an owned reference. Cython will automatically
-perform a ``Py_INCREF`` and ``Py_DECREF`` operation. Casting to
-``<PyObject *>`` creates a borrowed reference, leaving the refcount unchanged.
+         .. note:: Cython will not prevent a redundant cast, but emits a warning for it.
+
+        To get the address of some Python object, use a cast to a pointer type
+        like ``<void*>`` or ``<PyObject*>``.
+        You can also cast a C pointer back to a Python object reference
+        with ``<object>``, or a more specific builtin or extension type
+        (e.g. ``<MyExtType>ptr``). This will increase the reference count of
+        the object by one, i.e. the cast returns an owned reference.
+        Here is an example:
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+
+        .. literalinclude:: ../../examples/userguide/language_basics/casting_python.pxd
+            :caption: casting_python.pxd
+        .. literalinclude:: ../../examples/userguide/language_basics/casting_python.py
+            :caption: casting_python.py
+
+        Casting to ``cast(object, ...)`` creates an owned reference. Cython will automatically
+        perform a ``Py_INCREF`` and ``Py_DECREF`` operation. Casting to
+        ``cast(pointer(PyObject), ...)`` creates a borrowed reference, leaving the refcount unchanged.
+
+    .. group-tab:: Cython
+
+        .. literalinclude:: ../../examples/userguide/language_basics/casting_python.pyx
+            :caption: casting_python.pyx
+
+        The precedence of ``<...>`` is such that ``<type>a.b.c`` is interpreted as ``<type>(a.b.c)``.
+
+        Casting to ``<object>`` creates an owned reference. Cython will automatically
+        perform a ``Py_INCREF`` and ``Py_DECREF`` operation. Casting to
+        ``<PyObject *>`` creates a borrowed reference, leaving the refcount unchanged.
 
 .. _checked_type_casts:
 
 Checked Type Casts
 ------------------
 
-A cast like ``<MyExtensionType>x`` will cast x to the class
+A cast like ``<MyExtensionType>x`` or ``cast(MyExtensionType, x)`` will cast ``x`` to the class
 ``MyExtensionType`` without any checking at all.
 
-To have a cast checked, use the syntax like: ``<MyExtensionType?>x``.
+To have a cast checked, use the syntax like: ``<MyExtensionType?>x`` in Cython language
```

Comment:
Will be good to move to tabs

Reply:
Just as reminder, i wil not comment on every one like those

Reply:
Based on https://github.com/cython/cython/pull/4242#issuecomment-879666879 I would close it without change.